test/memory.cpp

```diff
@@ -796,7 +802,166 @@ void remove_memory_management_fn(af_memory_manager manager, int id) {}
+            manager, remove_memory_management_fn);
+
+        af_set_memory_manager(manager);
+    }
+
+    void TearDown() override {
+        af_device_gc();
+        af_unset_memory_manager();
+        af_release_memory_manager(manager);
+    }
+};
+
+TEST_F(MemoryManagerApi, E2ETest1D) {
+    size_t aSize = 8;
+
+    array a = af::array(aSize, af::dtype::f32);
+    ASSERT_EQ(payload->table.size(), 1);
+
+    ASSERT_EQ(payload->table[a.device<float>()], aSize * sizeof(float));
+    ASSERT_EQ(payload->lastNdims, 1);
+    ASSERT_EQ(payload->lastDims, af::dim4(aSize));
+    ASSERT_EQ(payload->lastElementSize, 4);
+}
+
+TEST_F(MemoryManagerApi, E2ETest2D) {
+    size_t aSize = 8;
+
+    af::array a = af::array(aSize, aSize, af::dtype::f32);
+    ASSERT_EQ(payload->table.size(), 1);
+    ASSERT_EQ(payload->table[a.device<float>()], aSize * aSize * sizeof(float));
+    ASSERT_EQ(payload->lastElementSize, 4);
+
+    // Currently this is set to 1 because all allocations request linear memory
+    // This behavior will change in the future
+    ASSERT_EQ(payload->lastNdims, 1);
+    ASSERT_EQ(payload->lastDims, af::dim4(aSize * aSize));
+}
+
+TEST_F(MemoryManagerApi, E2ETest3D) {
+    size_t aSize = 8;
+
+    af::array a = af::array(aSize, aSize, aSize, af::dtype::f32);
+    ASSERT_EQ(payload->table.size(), 1);
+    ASSERT_EQ(payload->table[a.device<float>()],
+              aSize * aSize * aSize * sizeof(float));
+    ASSERT_EQ(payload->lastElementSize, 4);
+
+    // Currently this is set to 1 because all allocations request linear memory
+    // This behavior will change in the future
+    ASSERT_EQ(payload->lastNdims, 1);
+    ASSERT_EQ(payload->lastDims, af::dim4(aSize * aSize * aSize));
+}
+
+TEST_F(MemoryManagerApi, E2ETest4D) {
+    size_t aSize = 8;
+
+    af::array a = af::array(aSize, aSize, aSize, aSize, af::dtype::f32);
+    ASSERT_EQ(payload->table.size(), 1);
+    ASSERT_EQ(payload->table[a.device<float>()],
+              aSize * aSize * aSize * aSize * sizeof(float));
+    ASSERT_EQ(payload->lastElementSize, 4);
+
+    // Currently this is set to 1 because all allocations request linear memory
+    // This behavior will change in the future
+    ASSERT_EQ(payload->lastNdims, 1);
+    ASSERT_EQ(payload->lastDims, af::dim4(aSize * aSize * aSize * aSize));
+    af::sync();
+}
+
+TEST_F(MemoryManagerApi, E2ETest4DComplexDouble) {
+    size_t aSize = 8;
+
+    af::array a = af::array(aSize, aSize, aSize, aSize, af::dtype::c64);
+    ASSERT_EQ(payload->table.size(), 1);
+    ASSERT_EQ(payload->table[a.device<float>()],
+              aSize * aSize * aSize * aSize * sizeof(double) * 2);
+    ASSERT_EQ(payload->lastElementSize, 16);
+
+    // Currently this is set to 1 because all allocations request linear memory
+    // This behavior will change in the future
+    ASSERT_EQ(payload->lastNdims, 1);
+    ASSERT_EQ(payload->lastDims, af::dim4(aSize * aSize * aSize * aSize));
+}
+
+TEST_F(MemoryManagerApi, E2ETestMultipleAllocations) {
```

Comment:
nit: Don't feel like the prefix `E2ETest` is adding anything to the readability factor..